Add tootctl email-domain-blocks

tateisu opened the pull request.

tateisu committed:
Add tootctl email_domains (block|unblock)

tateisu committed:
fix codeclimate issues.

tateisu committed:
fix codeclimate issues.

tateisu committed:
fix codeclimate issues.

tateisu committed:
add list subcommand, remove log_action.

tateisu commented:
tested on my test server. 
https://gist.github.com/tateisu/bce74cee1048b17082aa5cf3b019cb1e

tateisu committed:
fix codeclimate issues.

tateisu committed:
filter duplicate hostnames,ips before block

Gargron commented on lib/cli.rb:
```diff
@@ -53,6 +54,9 @@ def self.exit_on_failure?
     subcommand 'upgrade', Mastodon::UpgradeCLI
 
+    desc 'email_domains SUBCOMMAND ...ARGS', 'Manage E-mail domains'
+    subcommand 'email_domains', Mastodon::EmailDomainsCLI
```
I believe you can/should use `-` instead of `_` in CLI.

Also, while the command is called `email-domains`, the subcommands in it are only for listing and adding blocks, so it is a misnomer. I see the following options:

- Rename `list` to `list-blocks`
- Make `list` return *all* e-mail domains and add a `--blocked` option to list only blocked domains (expensive and probably useless, though). The command could also return nothing unless `--blocked` is passed (placeholder)
- Rename `email-domains` to `email-domain-blocks` and rename `block` to `add` and `unblock` to `remove`
- Integrate the subcommands with the `domains` command instead

tateisu replied:
I will change email_domains to email-domain-blocks.
But already `preview_cards` is exists.

tateisu replied:
I obey to your comment, but I figure out…

- Thor requires to use '_' instead of '-' in `subcommand` definition. when i use '-'  in the definition, the subcommand is not recognized. 
- Thor recognize both of '_' and '-' when command was invoked.
- Thor use '_'  in subcommand help.
- Thor use capitalization in subcommand `help`.

```
$ docker-compose run --rm web tootctl email-domain-blocks
Commands:
  tootctl email_domain_blocks add [DOMAIN...]     # add E-mail domain blocks
  tootctl email_domain_blocks help [COMMAND]      # Describe subcommands or one specific subcommand <=== THIS LINE
  tootctl email_domain_blocks list                # list E-mail domain blocks
  tootctl email_domain_blocks remove [DOMAIN...]  # remove E-mail domain blocks
```

Gargron commented on lib/mastodon/email_domains_cli.rb:
```diff
@@ -0,0 +1,107 @@
+    def list
+      EmailDomainBlock.where(parent_id: nil).includes(:children).order(id: 'DESC').find_each do |entry|
+        say(entry.domain.to_s, :green)
+        entry.children.order(id: 'DESC').find_each do |child|
```
If you execute SQL command `order` on `children` then `includes(:children)` will have no effect as the preloaded collections will not be used.

Gargron commented on lib/mastodon/email_domains_cli.rb:
```diff
@@ -0,0 +1,107 @@
+        email_domain_block = EmailDomainBlock.new(domain: domain, with_dns_records: options[:with_dns_records] || false)
+        email_domain_block.save!
+
+        say("#{domain} was blocked.", :green)
```
In other CLI commands, we just count up number of successes and then return a message at the end like "Blocked 123 domains"

Gargron commented on lib/mastodon/email_domains_cli.rb:
```diff
@@ -0,0 +1,107 @@
+      end
+
+      domains.each do |domain|
+        unless EmailDomainBlock.where(domain: domain).exists?
```
It would make more sense to use `find_by`, then warn if it's nil, then `destroy`, instead of doing `where(...).destroy_all` later. `destroy_all` might make sense if you pass an array of all the domains into `where`, but you are already doing `domains.each { ... }` so you are processing them one by one.

Gargron commented on lib/mastodon/email_domains_cli.rb:
```diff
@@ -0,0 +1,107 @@
+      end
+    end
+
+    desc 'unblock [DOMAIN...]', 'Unblock E-mail domains'
```
No capitalization within sentence.

Gargron commented on lib/mastodon/email_domains_cli.rb:
```diff
@@ -0,0 +1,107 @@
+    end
+
+    desc 'unblock [DOMAIN...]', 'Unblock E-mail domains'
+    long_desc <<-LONG_DESC
```
`long_desc` is not needed if it doesn't contain more information than `desc`.

Gargron commented on lib/mastodon/email_domains_cli.rb:
```diff
@@ -0,0 +1,107 @@
+    end
+
+    option :with_dns_records, type: :boolean
+    desc 'block [DOMAIN...]', 'Block E-mail domains'
```
No capitalization within sentence. Same in `long_desc` below

Gargron commented on lib/mastodon/email_domains_cli.rb:
```diff
@@ -0,0 +1,107 @@
+    LONG_DESC
+    def list
+      EmailDomainBlock.where(parent_id: nil).includes(:children).order(id: 'DESC').find_each do |entry|
+        say(entry.domain.to_s, :green)
```
I would not use green/yellow for this since they are not messages about the command. I think they should be white -- maybe blue.

tateisu committed:
rebase from currnet master branch.
rename email_domains_cli.rb to email_domain_blocks_cli.rb .
rename Mastodon::EmailDomainsCLI to Mastodon::EmailDomainBlocksCLI .
rename command email_domains to email-domain-blocks . (Thor recognizes both of - and _ )
rename subcommand block to add .
rename subcommand unblock to remove .
change the color in list subcommand to while for domain or cyan for childlen.
don't use include() in list subcommand.
suppress console output about succeeded entry.
add console output about count of processed/skipped.
remove capitalization in subcommand description.
remove long_desc in subcommand 'remove'.
remove duplicate where in subcommand 'remove'.

tateisu commented:
test log 2020/5/17 https://gist.github.com/tateisu/ba11d71cb15d3edcd768064cbdc254c0

tateisu committed:
fix codeclimate issue.
